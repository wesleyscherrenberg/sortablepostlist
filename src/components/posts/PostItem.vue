<script setup lang="ts">
import {computed} from 'vue';
import IconAngleUp from "@/components/icons/IconAngleUp.vue";
import IconAngleDown from "@/components/icons/IconAngleDown.vue";
import type {ActionsType, PostItemPropType} from "@/types/Types";

const emit = defineEmits(['onSortChange'])

const props = withDefaults(defineProps<PostItemPropType>(), {
  showSortUp: true,
  showSortDown: true,
});

const postTitle = computed(() => `Post ${props.postId}`)

const onSortHandler = (isAsc: boolean) => {
  const action: ActionsType = {
    postId: props.postId,
    fromIndex: props.index,
    toIndex: isAsc ? props.index + 1 : props.index - 1
  }

  emit('onSortChange', action)
}

</script>

<template>
    <div class="flex justify-center items-center shadow rounded bg-white min-h-[6rem] transform transition-all duration-500">
      <div class="flex-1 p-3 self-center" data-test="postTitle">{{ postTitle }}</div>
      <div class="flex flex-col justify-between" data-test="sortButtons" v-if="showSortUp || showSortDown">
          <button v-if="showSortUp"
                  aria-label="sortUp"
                  role="button"
                  data-test="sortUp"
                  class="p-3 transform transition-all duration-200 hover:scale-110"
                  @click="onSortHandler(false)">
            <icon-angle-up class="fill-theme-primary" />
          </button>

          <button v-if="showSortDown"
                  aria-label="sortDown"
                  role="button"
                  data-test="sortDown"
                  class="p-3 transform transition-all duration-200 hover:scale-110"
                  @click="onSortHandler(true)">
            <icon-angle-down class="fill-theme-primary" />
          </button>
      </div>
    </div>
</template>
