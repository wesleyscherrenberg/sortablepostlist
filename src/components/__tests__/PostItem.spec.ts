import { shallowMount, VueWrapper } from '@vue/test-utils';
import PostItem from "@/components/posts/PostItem.vue";
import {beforeEach, describe, expect, it} from "vitest";
import type {PostItemPropType} from "@/types/Types";

const factory = (props: PostItemPropType) => {
    return shallowMount(PostItem, {
        props
    });
}

const defaultProps: PostItemPropType = {
    postId: 1,
    index: 1,
    showSortUp: true,
    showSortDown: true,
}

describe('PostItem.vue', () => {
    let wrapper: VueWrapper;

    beforeEach(() => {
        wrapper = factory(defaultProps)
    });

    it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders the component', () => {
        expect(wrapper.exists()).toBe(true);
    });

    it('renders the postTitle correctly', () => {
        expect(wrapper.find('[data-test="postTitle"]').text()).toBe(`Post ${defaultProps.postId}`);
    });

    it('emits "onSortChange" event when sort buttons are clicked', async () => {
        await wrapper.find('[data-test="sortUp"]').trigger('click');
        expect(wrapper.emitted().onSortChange).toBeTruthy();
        expect(wrapper.emitted().onSortChange[0]).toEqual([{ postId: 1, fromIndex: 1, toIndex: 0 }]);

        await wrapper.find('[data-test="sortDown"]').trigger('click');
        expect(wrapper.emitted().onSortChange[1]).toEqual([{ postId: 1, fromIndex: 1, toIndex: 2 }]);
    });

    describe('when sort buttons', () => {

        it('should not render sortUp button when showSortUp is false', () => {

            wrapper = factory({...defaultProps, showSortUp: false});

            expect(wrapper.find('[data-test="sortUp"]').exists()).toBe(false);
        });

        it('should not render sortDown button when showSortDown is false', () => {

            wrapper = factory({...defaultProps, showSortDown: false});

            expect(wrapper.find('[data-test="sortDown"]').exists()).toBe(false);
        });

        it('should not render the element with sortUp and sortDown button when showSortUp and showSortDown is false', () => {

            wrapper = factory({...defaultProps, showSortDown: false, showSortUp: false});

            expect(wrapper.find('[data-test="sortButtons"]').exists()).toBe(false);
        });

    })
});
