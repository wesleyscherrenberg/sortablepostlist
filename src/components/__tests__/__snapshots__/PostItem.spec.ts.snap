// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostItem.vue > matches the snapshot 1`] = `
"<div class="flex justify-center items-center shadow rounded bg-white min-h-[6rem] transform transition-all duration-500">
  <div class="flex-1 p-3 self-center" data-test="postTitle">Post 1</div>
  <div class="flex flex-col justify-between" data-test="sortButtons"><button data-test="sortUp" class="p-3 transform transition-all duration-200 hover:scale-110">
      <icon-angle-up-stub class="fill-theme-primary"></icon-angle-up-stub>
    </button><button data-test="sortDown" class="p-3 transform transition-all duration-200 hover:scale-110">
      <icon-angle-down-stub class="fill-theme-primary"></icon-angle-down-stub>
    </button></div>
</div>"
`;
